Source/Core/InputCommon/ControllerInterface/evdev/evdev.cpp

```diff
@@ -291,121 +332,168 @@ ControlState evdevDevice::Button::GetState() const
+  case FF_SAW_UP:
+    return "Sawtooth Up";
+  case FF_SAW_DOWN:
+    return "Sawtooth Down";
   default:
-  {
-    const char* name = libevdev_event_code_get_name(EV_FF, m_type);
-    if (name)
-      return StripSpaces(name);
     return "Unknown";
   }
-  }
 }
 
-void evdevDevice::ForceFeedback::SetState(ControlState state)
+std::string evdevDevice::RumbleEffect::GetName() const
 {
-  // libevdev doesn't have nice helpers for forcefeedback
-  // we will use the file descriptors directly.
+  return m_use_strong_motor ? "Strong" : "Weak";
+}
 
-  if (m_id != -1)  // delete the previous effect (which also stops it)
+void evdevDevice::Effect::SetState(ControlState state)
+{
+  if (UpdateParameters(state))
   {
-    ioctl(m_fd, EVIOCRMFF, m_id);
-    m_id = -1;
+    // Update effect if parameters changed.
+    UpdateEffect();
   }
+}
 
-  if (state > 0)  // Upload and start an effect.
+void evdevDevice::Effect::UpdateEffect()
+{
+  // libevdev doesn't have nice helpers for forcefeedback
+  // we will use the file descriptors directly.
+
+  // Note: m_effect.type is set within UpdateParameters
+  // to determine if effect should be playing or not.
+  if (m_effect.type != DISABLED_EFFECT_TYPE)
   {
-    ff_effect effect;
-
-    effect.id = -1;
-    effect.direction = 0;        // down
-    effect.replay.length = 500;  // 500ms
-    effect.replay.delay = 0;
-    effect.trigger.button = 0;  // don't trigger on button press
-    effect.trigger.interval = 0;
-
-    // This is the the interface that XInput uses, with 2 motors of differing sizes/frequencies that
-    // are controlled seperatally
-    if (m_type == FF_RUMBLE)
-    {
-      effect.type = FF_RUMBLE;
-      // max ranges tuned to 'feel' similar in magnitude to triangle/sine on xbox360 controller
-      effect.u.rumble.strong_magnitude = u16(state * 0x4000);
-      effect.u.rumble.weak_magnitude = u16(state * 0xFFFF);
-    }
-    else  // FF_PERIODIC, a more generic interface.
+    if (-1 == m_effect.id)
     {
-      effect.type = FF_PERIODIC;
-      effect.u.periodic.waveform = m_type;
-      effect.u.periodic.phase = 0x7fff;  // 180 degrees
-      effect.u.periodic.offset = 0;
-      effect.u.periodic.period = 10;
-      effect.u.periodic.magnitude = s16(state * 0x7FFF);
-      effect.u.periodic.envelope.attack_length = 0;  // no attack
-      effect.u.periodic.envelope.attack_level = 0;
-      effect.u.periodic.envelope.fade_length = 0;
-      effect.u.periodic.envelope.fade_level = 0;
-    }
+      // If effect was not uploaded (previously stopped)
+      // we upload it and start playback
 
-    ioctl(m_fd, EVIOCSFF, &effect);
-    m_id = effect.id;
+      ioctl(m_fd, EVIOCSFF, &m_effect);
 
-    input_event play;
-    play.type = EV_FF;
-    play.code = m_id;
-    play.value = 1;
+      input_event play = {};
+      play.type = EV_FF;
+      play.code = m_effect.id;
+      play.value = 1;
 
-    if (write(m_fd, &play, sizeof(play)) < 0)
+      if (write(m_fd, &play, sizeof(play)) < 0)
```

Comment:
Empty control structures always bother me (since I don't know if it was intentionally blank or someone forgot to put something in there), would you mind either leaving a comment or rewriting it as something else (such as `static_cast<void>(write(...))` which seems to be the preferred way with `[[nodiscard]]` and friends)?

Reply:
I was copying the existing code-style. I assume it was to eliminate a warning for an ignored return value.

Reply:
Just noticed that (and that theres another place further up that also does this).
I'd appreciate if you changed it (since you're touching that part of the code anyways), but you're welcome to ignore me just as well ;)